### DISCUSSION TITLE
"await _" as replacement for "await Task.CompletedTask"

### DISCUSSION THREAD

**Discussion by datvm** [dotnet/csharplang#5457] [Language Ideas] [answered]
Sometimes a method signature requires an `async function` that we don't really need any `await`. Right now the placeholder would be `await Task.CompletedTask;` to dismiss compiler warning (I wonder if C# is smart enough to know we don't actually need to wait anything for this task).

My suggestion is to allow this: `await _` which signals that it's intended that this method doesn't need any `await` but we can't get rid of `async`. Compiler can also safely generate code that does not need any Task-related code.

Example:

```
// Library code:
public async Task DoSomethingAsync(Func<SomeType, Task> action);

// User
await lib.DoSomethingAsync(async instance => 
  {
      this.Update(instance);
      // No need any async here
     await _;
  })
```

**Comment by HaloFour** [comment]
> Sometimes a method signature requires an `async function` that we don't really need any `await`.  

Under what circumstance would `async` be required on a method signature?  That keyword is only necessary to enable the use of `await` within the method.  It does not change the method signature at all.  If you don't need to use `await` you should never need to use `async`.

**Reply by datvm** [reply]
Thanks, I updated the original post to add example scenario. For example, from library/framework signatures that you cannot change.

**Reply by HaloFour** [reply]
You're not required to mark the lambda as `async`.

```cs
// User
await lib.DoSomethingAsync(instance => 
  {
      // no async or await necessary here
      this.Update(instance);
      return Task.CompletedTask();
  });

**Reply by datvm** [reply]
You are. If you do not mark it, you get a compile error, since your lambda isn't technically returning a `Task`. Same for explicit functions.

![image](https://user-images.githubusercontent.com/6388546/142734717-7e466527-d276-43f2-91a5-c2c124c41190.png)

Or yes, like your case, you can return a `Task.CompletedTask`, but it's the same, it requires us to explicitly return it.

**Reply by HaloFour** [reply]
You do need to return a `Task` given that's the return type of the `Func<...>`, but that doesn't force the lambda to be `async`, just that you return some `Task`.  I did update my example.

**Reply by bernd5** [reply]
Sample:
```csharp
using System;
using System.Threading.Tasks;
using SomeType = System.String;


class App{
       
    public static async Task DoSomethingAsync(Func<SomeType, Task> action){
        await action("Foo");
    }
    
    public static async Task Main(){
         await DoSomethingAsync(s =>{
             Console.WriteLine($"Print {s}");
             return Task.CompletedTask; 
         });
    }
}
```

**Reply by datvm** [reply]
Yeah I concede that even with my proposal there is no line saved. I hoped there could be a shorter way to remove `Task.CompletedTask` completely. I assume `return _;` is too vague and C# wouldn't allow that.

**Reply by HaloFour** [reply]
I guess the question is how often does this situation arise that it warrants a special syntax.  Also, how would this work for `async` methods that don't return `Task`/`Task<T>`?

**Reply by datvm** [reply]
I see, admittedly I rarely worked with the methods you mentioned. You're right.

**Comment by jnm2** [comment, marked as answer]
@datvm I dismiss the compiler warning more directly, by silencing it in .editorconfig. I've never found it helpful. It even puts pressure towards an antipattern of doing stuff that might throw catchable exceptions before returning Task.CompletedTask.

**Reply by datvm** [reply]
Cool thanks for the idea about configuration. Can you clarify the bit at the end? I may be doing that without knowing.

**Reply by jnm2** [reply]
Sure. You might do this:

```cs
public Task DoSomethingAsync()
{
    DoSomethingElse();
    return Task.CompletedTask;
}
```

If now or in the future DoSomethingElse starts throwing an exception that makes sense to catch, let's say `IOException` or `OperationCanceledException`, then you might write code like this:

```cs
try
{
    await DoSomethingAsync();
}
catch (IOException)
{
    // Apply some policy at this level such as defaulting or retrying
}
```

So far so good. But things go terribly wrong if you don't instantly await:

```cs
var doSomethingTask = DoSomethingAsync(); // Program crashes here because DoSomethingAsync didn't use `async`

// Do something else

try
{
    await doSomethingTask;
}
catch (IOException)
{
    // This catch block can only catch exceptions that were returned in a Task value from DoSomethingAsync
}
```

In other words, the place to expect exceptions has become unpredictable and you basically need two try...catches or to overly expand the scope of the catch.

It becomes brittle in the face of future refactorings that will appear benign.

**Reply by datvm** [reply]
Goodness! I totally didn't believe it could happen so I had to try the code by myself.

Can confirm this works correctly although there is warning

```c#
var t = DoSomethingAsync();

try
{
    await t;
}
catch (IOException)
{
    Console.WriteLine("Caught");
}

static async Task DoSomethingAsync()
{
    DoSomething();
    //return Task.CompletedTask;
}

static void DoSomething()
{
    throw new IOException();
}
```

Removing `async` and using `return Task.CompletedTask` instead would result in uncaught exception at the first line.

**Reply by Kobus-Smit** [reply]
@datvm, @jnm2 But using `await Task.CompletedTask` instead of `return Task.CompletedTask` does result in predictable exceptions.


```csharp
var t = DoSomethingAsync();

try
{
    await t;
}
catch (IOException)
{
    Console.WriteLine("Caught");
}

static async Task DoSomethingAsync()
{
    DoSomething();
    await Task.CompletedTask;
}

static void DoSomething()
{
    throw new IOException();
}
```

**Reply by jnm2** [reply]
@Kobus-Smit Exactly, but I'd much rather say what I really mean (disable the CS1998 warning) than add a no-op statement that comes with runtime overhead and isn't as direct about why it is there.

**Reply by Kobus-Smit** [reply]
@jnm2 Yeah true, but disabling the CS1998 feels wrong too, because sometimes CS1998 is a relevant warning. I'm not always sure which one is the lesser of the two evils...

**Reply by jnm2** [reply]
@Kobus-Smit I'm still waiting for the day when that warning tells me something I wasn't going to see anyway. The fact that it used to sometimes beat me to seeing it apparently did not mean that I wasn't going to notice it at all. And how terrible would it be to miss noticing it? I'm already okay with async methods that don't have awaits in general as explained above.

**Reply by TahirAhmadov** [reply]
@jnm2 disabling that warning is a bad idea. It tells you that a method is `async` where it doesn't need to be. The state machine overhead is not free, not even in latest versions with allocation-free structs. I measured performance and simple string operation (JSON output) becomes 50% slower due to the state machine overhead (when the actual output just goes to memory and actual IO is not involved). When IO is involved, then these nano-seconds don't matter - giving thread back is more valuable - but to dismiss this overhead wholesale is imprudent.

**Reply by jnm2** [reply]
That overhead is worth it for the correctness. To me, this is a good tradeoff. *Any* async method can be made non‐async and hyperoptimized. Optimization is an option as usual where it is shown to be worth it.

**Reply by TahirAhmadov** [reply]
There is nothing saying that it is **correct** to mark synchronous `Task` returning methods as `async`. There is no such rule, best practice guidance, or any valid reason to do so.
Furthermore, when you have a `Task` returning method which only calls one other `Task` returning method, and that call is the last statement in the method, I haven't seen any good reason to do `async` and `await` - it's better to just return the `Task` as the return value, to reduce overhead.

> Optimization is an option as usual where it is shown to be worth it.

Wrong. Always write highest performance code. The only two considerations more important than performance are security and reliability. Making synchronous `Task` returning methods do the state machine overhead does not improve neither security nor reliability.

**Reply by CyrusNajmabadi** [reply]
> Wrong. Always write highest performance code. 

That's patently ridiculous. I wouldn't be able to even use c# of that was the rule to follow. I couldn't use any types in the bcl. I would likely be spending offers of magnitude more time and energy getting a tiny bit faster in ways that wouldn't impact the end result meaningfully for consumers. 

Write the code that allows you to efficiently use your time to solve the problems you have. Focus your time and energy where relevant to your domain.

**Reply by theunrepentantgeek** [reply]
I'm not sure of the exact phrasing or origin, but I've been following this advice for a couple of decades:

> First make it work.
> Then make it understandable.
> Then make it fast, if needed.
> In strictly that order.

A [close variant](https://wiki.c2.com/?MakeItWorkMakeItRightMakeItFast) of this phrase is attributed to Kent Beck.

I have no doubt that there are times when wringing the absolute best performance out of a piece of code is required. I've found those times are startlingly rare - and most often accommodated by using a better algorithm.

**Reply by jnm2** [reply]
I've provided valid reasons for this above, and explained the correctness pitfall with eliding async/await. In other threads I've listed even more reasons (tooling loses stack trace) and mentioned Stephen Cleary's recommendation on this also.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi that's not what I meant. First, with most BCL types, it's unlikely most developers can write their own version which will be faster. `List<T>` does what it's supposed to, there isn't much to optimize there. Etc. Second, if it doesn't impact the consumers, obviously it's not faster. However, even small optimizations add up over time. This is especially important with custom library code which is used in many places. Third, following best practices, which come at no significant extra development cost, and give you even a slight performance improvement, is better than doing something else which takes a similar amount of effort but is slower, even if slightly. Case in point - `async` synchronous methods with warning suppression, vs. just returning `Task.CompletedTask`. It's plus or minus a couple of words, the same amount of effort, but one option is slightly faster. Why go with the slower version?

In short, I continue being surprised by the cavalier attitude many developers take towards performance. How many CPU gigahertz and physical cores, gigabytes of RAM, and watts of power consumption will be needed to run Notepad, (figuratively speaking) before this attitude changes?

**Reply by TahirAhmadov** [reply]
> > First make it work.
> > Then make it understandable.
> > Then make it fast, if needed.
> > In strictly that order.

Respectfully, I disagree. If by "understandable" it's referring to readable and maintainable code, that's definitely a consideration, but it's below performance. The customer wants working, fast software; he doesn't care about your life being more or less difficult. This prioritization of non-customer concern over customer concern is fundamentally flawed, IMO.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> Respectfully, I disagree. If by "understandable" it's referring to readable and maintainable code, that's definitely a consideration, but it's below performance.

Respectfully, I disagree as well.  If it's not understandable then it's not maintainable, and if it's not maintainable then it doesn't really matter how fast it is.  Considerations for performance should be made but only where it's worth the complexity, and even then very carefully and very well documented.  One must consider that it's not just other developers who have to be able to read this code, it's yourself six months or more from now.

**Reply by TahirAhmadov** [reply]
> If it's not understandable then it's not maintainable, and if it's not maintainable then it doesn't really matter how fast it is. 

Doesn't matter - to whom? The customer wants working and fast software today and in 6 months. If you deliver slow software, the customer is very unlikely to be receptive to the software provider's arguments that "hey it's slow, but believe us - the code is well structured and easy to maintain!" The customer will listen to it, smile, maybe even say something positive, and then buy a different product. (Leaving corporate procurement shenanigans aside.)

> Considerations for performance should be made but only where it's worth the complexity, and even then very carefully and very well documented. One must consider that it's not just other developers who have to be able to read this code, it's yourself six months or more from now.

Again, I'm not advocating for squeezing out performance at the cost of extreme complexity. The specific case we're discussing here - `async` vs `return Task.CompletedTask` - the complexity is almost equal. I still haven't heard anybody explain why it's so important to do `async` and sacrifice all those CPU cycles.

**Reply by CyrusNajmabadi** [reply]
> s not what I meant. First, with most BCL types, it's unlikely most developers can write their own version which will be faster. 

You definitely can. bcl types are designed to be flexible and good for a very broad audience, often at the cost of performance. For most domains you can do better since you don't need that. 

> List<T> does what it's supposed to, there isn't much to optimize there.

Again not true :-)

In practice though optimizing this will give extremely marginal wins. If it's your hot spot, it can be worth it. If not, don't bother. 

> Why go with the slower version?

@jnm2 already explained why. It actually has the semantics he wants. For example async exceptions, not sync ones. 

> In short, I continue being surprised by the cavalier attitude many developers take towards performance.

I've already explained why. We would get nowhere with your attitude. I personally would have to spend orders of magnitude more time per feature for gains that would not be relevant to my users. That's extremely wasteful and honestly pointless.

**Reply by CyrusNajmabadi** [reply]
> Doesn't matter - to whom? The customer wants working and fast software today and in 6 months. If you deliver slow software, 

That's the point though. In practice this is unlikely to matter. And if it does matter, then that's what you optimize. That's that everyone is saying. Spend your limited time and resource budget on things that end up actually mattering.

**Reply by CyrusNajmabadi** [reply]
> and sacrifice all those CPU cycles.

Again, this sacrifice is only a problem if it impacts someone in a noticeable way. It's like railing against virtual methods, or generic method dispatch. Yes, they have overhead, but if that overhead doesn't substantively affect the end program, then it's not an issue and it's reasonable to do if it expresses the semantics you want and allows you to expend your energy solving harder problems.

**Reply by TahirAhmadov** [reply]
> You definitely can. bcl types are designed to be flexible and good for a very broad audience, often at the cost of performance. For most domains you can do better since you don't need that.

I didn't run into this so far. I never needed a collection which has semantics of `List<T>` but can be written in a way that's faster than the way existing `List<T>` does it.
If a BCL type can be optimized, great - that can be submitted to BCL, I'm sure they'll appreciate it. Again, I don't recall ever needing to do that.

> @jnm2 already explained why. It actually has the semantics he wants. For example async exceptions, not sync ones.

For most use cases, if the ultimate caller does `await` and that call is in a `try/catch`, does it really matter how the exception is raised?

**Reply by TahirAhmadov** [reply]
> > and sacrifice all those CPU cycles.
> 
> Again, this sacrifice is only a problem if it impacts someone in a noticeable way. It's like railing against virtual methods, or generic method dispatch. Yes, they have overhead, but if that overhead doesn't substantively affect the end program, then it's not an issue and it's reasonable to do if it expresses the semantics you want and allows you to expend your energy solving harder problems.

A lot of my work these days is web services - user facing and API - and I'm constantly looking for ways to ensure that under heavy load, the web server is not experiencing unreasonable CPU/memory usage. I used to follow this approach of "get it to work, worry about performance later", until I saw the result - production spikes with actual impact on users. 
To summarize, my point is, there are a few simple rules to write performance and scalable code without adding extra complexity and re-engineering BCL, and following those rules would make a big difference across the industry.

**Reply by CyrusNajmabadi** [reply]
> I never needed a collection which has semantics of List<T> but can be written in a way that's faster

And @jnm2 hasn't needed a method with the semantics of an async method, but which needs to be faster.  So he picks the form that expresses what he wants and is fast enough. 

> If a BCL type can be optimized, great - that can be submitted to BCL, 

It can't be. The general design of List is to be flexible. If you make it better in one way, it might hurt another.  So if you need it faster for the things you care about, you need to roll your own or go elsewhere. 

> Again, I don't recall ever needing to do that.

Exactly. For *your needs* it's sufficient. Even though you're leaving perf on the table all over the place. That's what people are saying to you: you pick the things that work and are understandable. And *then* you spend efforts on perf only where suitable.

**Reply by CyrusNajmabadi** [reply]
> For most use cases, 

We don't follow this pattern in much of our code. Heck, we still have tons of code that still uses manual continuations and extension methods on Task as we find that expresses our intent better than async/await. Having two forms of exception paths mucks this up. If the perf benefit is not at all relevant to that particular domain you've now made things much more complex and you can introduce problems (this is not a hypothetical). This has impacted us.

**Reply by CyrusNajmabadi** [reply]
> and following those rules would make a big difference across the industry.

You are talking your domain and applying it to everyone else. Furthermore, you are not even consistently following your own advice.  You said:

> Wrong. Always write highest performance code.

But you don't follow that because even you have realized that the extra perf you eek out from this strategy is so minor as to not be worth it in nearly all cases. 

You are practicing the philosophy everyone else is mentioning. You accept the perf trade-offs, and you measure and you focus your efforts on the parts that actually matter to your domain. Please accept that that is what other people do as well, and on their domains the perf trade-offs may be entirely different.

**Reply by jnm2** [reply]
> For most use cases, if the ultimate caller does await and that call is in a try/catch, does it really matter how the exception is raised?

@TahirAhmadov It would be a mistake to assume this is the only way that task-returning methods are used. The amount of times that `await` is not immediately used *is* the pitfall I was talking about when I made the point above. People like to parallelize by not immediately awaiting; I've seen this (and antipatterns of this) often. It comes up in real apps and in open-source libraries that handle tasks and task-returning delegates. People also sometimes examine IsCompleted before awaiting to provide different UI behavior.

You're making statements like this:

> There is nothing saying that it is correct to mark synchronous Task returning methods as async. There is no such rule, best practice guidance, or any valid reason to do so.

Which demonstrates that if you've read my comments that you're replying to, you've dismissed my preexisting answers to this as invalid without providing any reason why. This is not a friendly conversational move and it doesn't give me any confidence that repeating myself (what else can I do in response?) will actually move the conversation forward.

**Reply by TahirAhmadov** [reply]
@jnm2 in your catching exception example, if you control both sides of the equation - the caller and all the possible callees - yes you can adopt the approach of `async` everything and you get the benefit of being able to await tasks after creating them in a try-catch. However, given that there is no language rule that all task-returning methods have to be async, it's possible some callees do not follow your guideline - specifically, those from third-party code (if you use any at all). This goes to the root of "abstracting" logic into separate methods, classes, libraries, etc. - your approach implies that there is a behavior expected from a method which is not guaranteed by its signature. As such, I don't think it's a good idea. If you want to await your task after its created, I think the correct solution is to try-catch both the creation and the await; you can do it within the same try-catch or different ones, as needed. If you don't want to copy-pasta try catch because it handles multiple exception types, you can put that into a reusable method:
```cs
void HandleException(Exception ex)
{
...
}
Task task;
try { task = DoSomethingAsync(); } catch(Exception ex) { HandleException(ex); }
// other code ....
try { await task; } catch(Exception ex) { HandleException(ex); }
```
Yes the above is a little more awkward, but it's in agreement with how `async` was implemented - in that it's optional.
This raises another question - given this use case, does it make sense to add something to the method signature (including delegates, interface and class methods) to "promise" that it's an `async` method, and only `await` throws? Basically:
```cs
interface IFoo { async required Task Bar(); }
class Foo : IFoo
{
  public async required Task Bar() // no error; no warning about lack of awaits
  {
    DoSomethingElse(); 
  }
  public async Task Bar() // no error; warning about lack of awaits
  {
    DoSomethingElse(); 
  }
  public Task Bar() // error
  {
    DoSomethingElse(); 
    return Task.CompletedTask;
  }
}

IFoo foo = ...;
var task = foo.Bar(); // this is guaranteed not to throw by method signature
try
{
  await task; // here we catch stuff in one spot
}
catch(...) { ... }
```

**Reply by CyrusNajmabadi** [reply]
> Yes the above is a little more awkward,

This feels like a lose lose. The code is more complex and more awkward and brings no value to the domain the person is operating in. 

Taking the List example from above. We could likely make all your code slightly faster by just using an array anda count. It would often be faster (and better on the GC with only one alloc instead of two), but it would be 'more awkward' and just a pointless waste for the vast majority of use cases. Yes, you might want to do it where measurements showed an appreciable benefit to a particular problem space. But absent that if it's correct and clear, then it's sufficient. 

Again, this is the philosophy you yourself engage in for other cases. You are just not willing to extend the same courtesy for others for this particular case.

**Reply by TahirAhmadov** [reply]
> You are talking your domain and applying it to everyone else.

I agree that performance is more important in some areas than others. However, I doubt there is a domain where it doesn't matter at all.

> Furthermore, you are not even consistently following your own advice. You said: ....

Maybe I didn't clearly state what I meant. When I said "highest performance", obviously it comes with the caveats that it should satisfy the requirements, be secure, reliable, etc. Also, I fully recognize the concept of iteration - you want to deliver working but slow code today, and will improve performance in 2 weeks; that's perfectly fine in the real world. 
My biggest issue is with the messaging of "performance doesn't matter, unless you are in one of those domains where it does" and "use this style which foregoes performance from the get-go".

**Reply by jnm2** [reply]
It's never "performance doesn't matter." It's "performance is acceptable so we schedule our time accordingly."

**Reply by CyrusNajmabadi** [reply]
@TahirAhmadov 

> in your catching exception example, if you control both sides of the equation - the caller and all the possible callees - yes you can adopt

This is all that is relevant. For any particular project the person who owns that project determines what is relevant for them. You do not get to dictate what they must find important or optimize for. As a customer you can request it, but the owner has ultimately the one who makes the determination on the balance of an their concerns. 

As you have demonstrated, you balance things the same way. You leave perf on the table when you want when you realize that it's just not worth it to you for your code to optimize to that level. That is not only understandable and acceptance, it's also appropriate. The most valuable resource we have is our own time and mental energy we have to take in our problem domains. Your insistence that everyone allocate that scarse resource as you would is not only inappropriate but actually rude. It presumes your views here are the only right ones and that everyone must agree and follow them to be in the right. 

For your problems you get to decide your hierarchy of importance. Please give others the courtesy to do the same for theirs.

**Reply by TahirAhmadov** [reply]
> This feels like a lose lose. The code is more complex and more awkward and brings no value to the domain the person is operating in.

More importantly, through, it works off the way async is implemented, and off what method signature promises.

> Taking the List example from above. We could likely make all your code slightly faster by just using an array anda count. It would often be faster (and better on the GC with only one alloc instead of two), but it would be 'more awkward' and just a pointless waste for the vast majority of use cases. Yes, you might want to do it where measurements showed an appreciable benefit to a particular problem space. But absent that if it's correct and clear, then it's sufficient.

1, that is impossible when the list has to be exposed as a property/field/etc. 2, even in locals scenario, if you go around re-implementing this stuff, the possibility of bugs goes up significantly, so it goes against the other thing I mentioned - reliability is more important than performance.

**Reply by CyrusNajmabadi** [reply]
> However, I doubt there is a domain where it doesn't matter at all.

No one's said this at all. Indeed everyone says the opposite. That performance makes total sense to focus on when relevant. You are the only one who had the absolutist position of:

> Wrong. Always write highest performance code.

You don't follow that position yourself, but you also claim others are wrong to make a determination for their domain if the perf cost of going this route is acceptable. As per my prior post this is pretty dang rude.

**Reply by CyrusNajmabadi** [reply]
> you want to deliver working but slow code today, and will improve performance in 2 weeks; that's perfectly fine in the real world.

It's also fine if you never improve the perf on it. There are parts of our code base that maybe run in 10ns , but I could get it to 1ns. That code runs once. I will never ever optimize this. It's utterly pointless. And, in fact, it would be doing my customers a disservice by not focusing on actual performance issues they are facing in the real world *all of which* are due to much more important issues than using `async` on a synchronous method.

**Reply by TahirAhmadov** [reply]
> > in your catching exception example, if you control both sides of the equation - the caller and all the possible callees - yes you can adopt
> 
> This is all that is relevant. For any particular project the person who owns that project determines what is relevant for them. 

1, I'm not demanding @jnm2 changes his coding style; thankfully, we live in a free society - he can do whatever he wants. 2, I'm voicing my disagreement with his approach as constructive criticism only because it came up on my radar, having been posted here. I don't see myself spending time policing anybody. 3, my biggest issue is the advice given to others: he didn't say "in my situation, I often await tasks after creation, and it's important to me to only throw on await"; he said (paraphrasing) "I just disable that warning wholesale and you should, too". Just like you are asking me to respect his choices, I'm only asking that those choices are not presented as a good practice across the board. (Especially when frankly I think it's not even a good practice in that case.)

**Reply by CyrusNajmabadi** [reply]
> More importantly, through, it works off the way async is implemented, and off what method signature promises.

That is up to the code owner to decide on relevance. If I return an int, and I know it's always positive, it's not for you to say that my code is wrong for not handling negative cases on the return side.

**Reply by TahirAhmadov** [reply]
> It's also fine if you never improve the perf on it. 

Oh yes. Obviously, if you can spend your week getting 2 seconds, or 10 ns, you want to hit the 2 seconds first. Again, partially because worded my stance a bit too absolutely, and partially because you read into it that I'm "fanatically" behind that stance, there seems to be disagreement where there isn't really.

**Reply by CyrusNajmabadi** [reply]
> m voicing my disagreement with his approach as constructive criticism only because it came up on my radar, 

You told him he was wrong and that he had to do something a particular way. Something which you yourself do not do when it suits you. It's fine to say: hey, just be aware that there is overhead here, and determine if that overhead is ok for your domain. 

Your posts treat the other person in a belittling fashion. You indicate that they are not capable of making this determination for their domain and that they are "wrong" of they don't follow this route. That's not ok.

**Reply by TahirAhmadov** [reply]
> > More importantly, through, it works off the way async is implemented, and off what method signature promises.
> 
> That is up to the code owner to decide on relevance. If I return an int, and I know it's always positive, it's not for you to say that my code is wrong for not handling negative cases on the return side.

Hmm, I'll probably disagree on that. I know using `uint` across the board was difficult due to unrelated reasons, but I really really wish those could've been overcome. Also, somebody posted a proposal for "method contracts" - I like that idea, as well. Yes, `int` is weird, because of CPU architecture and what not is the default type for all integers, and while I reluctantly go with the flow here (due to lack of options), I would not extend the same logic to other types.

**Reply by CyrusNajmabadi** [reply]
> Oh yes. Obviously, if you can spend your week getting 2 seconds, or 10 ns, you want to hit the 2 seconds first. 

That's not my actual point there. . Say I have no big perf issues. I may never ever ever try to make that 10ns hit better. *ever*. And I'm not wrong for doing that. Max perf is not a goal in and of itself. Delivering a product that meets the perf needs of the customers is the goal. After that anything else is optional.  And that means potentially *never* doing anything there.

**Reply by CyrusNajmabadi** [reply]
> Hmm, I'll probably disagree on that.

Tough :-)

Me methods that return the lengths of things will continue to use int, and my consumption code will continue to ignore negatives. There is nothing 'right' or 'wrong' about this. This is a call I make as the code owner in terms of what I think should get attention and what is irrelevant.

**Reply by TahirAhmadov** [reply]
> > Oh yes. Obviously, if you can spend your week getting 2 seconds, or 10 ns, you want to hit the 2 seconds first.
> 
> That's not my actual point there. . Say I have no big perf issues. I may never ever ever try to make that 10ns hit better. _ever_. And I'm not wrong for doing that. Max perf is not a goal in and of itself. Delivering a product that meets the perf needs of the customers is the goal. After that anything else is optional. And that means potentially _never_ doing anything there.

Sure. If in your case, you have this one code path, it takes 10 ns, is only hit once, and no human can ever notice it, it's not a performance problem. That's why I clarified - it adds up with repetitive invocations and in reusable library code.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi what do you think about my `async required` proposal though? Isn't that the solution to the problem? It will accommodate @jnm2 nicely, and is an explicit behavior which communicated the intention clearly.

**Reply by CyrusNajmabadi** [reply]
> Again, partially because worded my stance a bit too absolutely, and partially because you read into it that I'm "fanatically" behind that stance

That's what telling someone they're wrong will do :-)

Notice that you have given reasons why it's ok to have overhead in your own program with bcl types. You've realized that there can be justifications for not having to take the position you put forth. But you still have not accepted that other people are adults with skills and knowledge on the topics and they can make their own assessments as well and that is ok with those assessments to diverge from yours. Even above you will not let me have my own assessment that I don't need to check for negative cases from methods I know can not return them. Nor can I use 'int' even if I think that is the most suitable type here.

**Reply by CyrusNajmabadi** [reply]
> what do you think about my async required proposal though? Isn't that the solution to the problem? It will accommodate @jnm2 nicely, and is an explicit behavior which communicated the intention clearly.

Generally it seems minor enough that I don't think we'd go for it.  But who knows, if it got a lot of interest, then maybe. You seem to be in a very particular realm though where this overhead is significant for you. That automatically puts you in the vast vast minority of users. But if we had some major libraries wanting this... Perhaps

**Reply by CyrusNajmabadi** [reply]
> Sure. If in your case, you have this one code path, it takes 10 ns, is only hit once, and no human can ever notice it, it's not a performance problem. 

Right. So jnm is not "wrong". He is the person that can make that assessment for his code. 

> he said (paraphrasing) "I just disable that warning wholesale and you should, too". 

He didn't say that. He said *not paraphrasing*:

> @datvm I dismiss the compiler warning more directly, by silencing it in .editorconfig. I've never found it helpful. It even puts pressure towards an antipattern of doing stuff that might throw catchable exceptions before returning Task.CompletedTask.

And

> To me, this is a good tradeoff. Any async method can be made non‐async and hyperoptimized. Optimization is an option as usual where it is shown to be worth it.

In ask these cases he has said what makes sense for him, and he absolutely backs out up with the understanding and good advice that if data shows there is an issue that it would make sense to optimize. 

You are attacking a position not put forth. And you have eventually just pivoted to the position clearly stated which you were originally attacking.

**Reply by TahirAhmadov** [reply]
> > @datvm I dismiss the compiler warning more directly, by silencing it in .editorconfig. I've never found it helpful. It even puts pressure towards an antipattern of doing stuff that might throw catchable exceptions before returning Task.CompletedTask.
> 
> > To me, this is a good tradeoff. Any async method can be made non‐async and hyperoptimized. Optimization is an option as usual where it is shown to be worth it.

I didn't notice the second comment. I don't know, to me, it sounded very done and shut when I first read it. Perhaps I mis-interpreted it; I'm sensitive to this topic. Extremely annoyed by slow websites, desktop apps, etc. Zero patience left for this stuff.

**Reply by TahirAhmadov** [reply]
I created https://github.com/dotnet/csharplang/discussions/5957 for everybody's consideration.
See, @jnm2, I care about you - wrote that up to make your life easier. :)

**Reply by jnm2** [reply]
Patience is exactly what is needed to reach worthwhile conclusions in these situations.
